File: recsystem-frontend/src/utils/authErrors.js
export function parseLoginError(err) {
  const raw = err?.response?.data ?? err?.message ?? ''
  const text = typeof raw === 'string' ? raw : String(raw)

  if (
    text.includes('Неверный логин или пароль') ||
    text.includes('invalid credentials') ||
    text.includes('wrong arguments') ||
    text.includes('rpc error')
  ) {
    return 'Неверный логин или пароль'
  }

  if (err?.response?.status === 401) {
    return 'Неверный логин или пароль'
  }

  return 'Не удалось войти. Попробуйте позже'
}

export function parseRegisterError(err) {
  const raw = err?.response?.data ?? err?.message ?? ''
  const text = typeof raw === 'string' ? raw : String(raw)

  if (text.includes('уже зарегистрирован') || text.includes('already exists')) {
    return 'Пользователь с таким email уже зарегистрирован'
  }

  if (text.includes('Проверьте правильность') || text.includes('wrong arguments')) {
    return 'Проверьте правильность заполнения полей (имя и фамилия — латиницей, пароль от 6 символов)'
  }

  return 'Не удалось зарегистрироваться. Попробуйте позже'
}

File: api-gateway/internal/http/handlers/auth_errors.go
package handlers

import (
	"net/http"
	"strings"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

func writePlainError(w http.ResponseWriter, status int, message string) {
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	w.WriteHeader(status)
	_, _ = w.Write([]byte(message))
}

func respondLoginError(w http.ResponseWriter, err error) {
	writePlainError(w, http.StatusUnauthorized, "Неверный логин или пароль")
}

func respondRegisterError(w http.ResponseWriter, err error) {
	if st, ok := status.FromError(err); ok {
		switch st.Code() {
		case codes.AlreadyExists:
			writePlainError(w, http.StatusConflict, "Пользователь с таким email уже зарегистрирован")
			return
		case codes.InvalidArgument:
			writePlainError(w, http.StatusBadRequest, "Проверьте правильность заполнения полей")
			return
		}
	}

	if strings.Contains(strings.ToLower(err.Error()), "already exists") {
		writePlainError(w, http.StatusConflict, "Пользователь с таким email уже зарегистрирован")
		return
	}

	writePlainError(w, http.StatusInternalServerError, "Не удалось зарегистрироваться. Попробуйте позже")
}
